framework/test/jest/unit/specs/modules/chain/dpos/apply.spec.js

```diff
@@ -0,0 +1,466 @@
+				{
+					fees: totalFee, // dividable to ACTIVE_DELEGATE count
+					rewards: delegatesWhoForged.map(() => rewardPerDelegate),
+					delegates: delegatesWhoForged.map(account => account.publicKey),
+				},
+			]);
+		});
+
+		it('should increase "missedBlocks" field by "1" for the delegates who missed forging slot', async () => {
+			// Act
+			await dpos.apply(lastBlockOfTheRound, stubs.tx);
+
+			// Assert
+			expect(
+				stubs.storage.entities.Account.increaseFieldBy,
+			).toHaveBeenCalledWith(
+				{
+					publicKey_in: delegatesWhoMissed.map(a => a.publicKey),
+				},
+				'missedBlocks',
+				'1',
+				stubs.tx,
+			);
+		});
+
+		it('should distribute rewards and fees ONLY to the delegates who forged', async () => {
+			// Act
+			await dpos.apply(lastBlockOfTheRound, stubs.tx);
+
+			// Assert
+			expect.assertions(constants.ACTIVE_DELEGATES);
+
+			// Assert Group 1/2
+			uniqueDelegatesWhoForged.forEach(account => {
+				expect(stubs.storage.entities.Account.update).toHaveBeenCalledWith(
+					{
+						publicKey_eq: account.publicKey,
+					},
+					expect.any(Object),
+					stubs.tx,
+				);
+			});
+
+			// Assert Group 2/2
+			delegatesWhoMissed.forEach(account => {
+				expect(stubs.storage.entities.Account.update).not.toHaveBeenCalledWith({
+					publicKey_eq: account.publicKey,
+				});
+			});
+		});
+
+		it('should distribute more rewards and fees (with correct balance) to delegates based on number of blocks they forged', async () => {
+			// Act
+			await dpos.apply(lastBlockOfTheRound, stubs.tx);
+
+			// Assert
+			expect.assertions(uniqueDelegatesWhoForged.length);
+			uniqueDelegatesWhoForged.forEach(account => {
+				const { fee, reward } = getTotalEarningsOfDelegate(account);
+				const amount = reward + fee;
+				const data = {
+					...account,
+					balance: account.balance.plus(amount),
+					fees: account.fees.plus(fee),
+					rewards: account.rewards.plus(reward),
+				};
+
+				expect(stubs.storage.entities.Account.update).toHaveBeenCalledWith(
+					{
+						publicKey_eq: account.publicKey,
+					},
+					data,
+					stubs.tx,
+				);
+			});
+		});
+
+		it('should give the remainingFee ONLY to the last delegate of the round who forged', async () => {
+			// Arrange
+			const remainingFee = randomInt(5, 10);
+			stubs.storage.entities.Round.summedRound.mockResolvedValue([
+				{
+					fees: totalFee + remainingFee,
+					rewards: delegatesWhoForged.map(() => rewardPerDelegate),
+					delegates: delegatesWhoForged.map(a => a.publicKey),
+				},
+			]);
+
+			// Act
+			await dpos.apply(lastBlockOfTheRound, stubs.tx);
+
+			// Assert
+			expect.assertions(uniqueDelegatesWhoForged);
+			expect(stubs.storage.entities.Account.update).toHaveBeenCalledWith(
+				{
+					publicKey_eq: delegateWhoForgedLast.publicKey,
+				},
+				expect.objectContaining({
+					/**
+					 * Delegate who forged last also forged 3 times,
+					 * Thus will get fee 3 times too.
+					 */
+					fees: delegateWhoForgedLast.fees.add(
```

Comment:
This comment is not true based on the test data provided, I think. The first delegateForged thrice and delegateWhoForgedLast is the last delegate who forged any block. Maybe double check with the data?

Reply:
I think it's corrected

```js
// framework/test/jest/unit/specs/modules/chain/dpos/round_delegates.js
const delegatesWhoForged = [
	...delegatesWhoForgedThrice,
	...delegatesWhoForgedTwice,
	...delegatesWhoForgedOnce,
	...delegatesWhoForgedTwice,
	...delegatesWhoForgedThrice,
	...delegatesWhoForgedThrice,
];

const delegateWhoForgedLast = delegatesWhoForged[delegatesWhoForged.length - 1];
```

Reply:
You are right!